rust/agama-lib/src/product/http_client.rs

```diff
@@ -0,0 +1,75 @@
+use crate::software::model::RegistrationInfo;
+use crate::software::model::RegistrationParams;
+use crate::software::model::SoftwareConfig;
+use crate::{base_http_client::BaseHTTPClient, error::ServiceError};
+
+pub struct ProductHTTPClient {
+    client: BaseHTTPClient,
+}
+
+impl ProductHTTPClient {
+    pub fn new() -> Result<Self, ServiceError> {
+        Ok(Self {
+            client: BaseHTTPClient::new()?,
+        })
+    }
+
+    pub fn new_with_base(base: BaseHTTPClient) -> Self {
+        Self { client: base }
+    }
+
+    // FIXME get_software_config ?
+    pub async fn get_config(&self) -> Result<SoftwareConfig, ServiceError> {
+        self.client.get("/software/config").await
+    }
+
+    pub async fn set_config(&self, config: &SoftwareConfig) -> Result<(), ServiceError> {
+        self.client.put_void("/software/config", config).await
+    }
+
+    /// Returns the id of the selected product to install
+    pub async fn product(&self) -> Result<String, ServiceError> {
+        let config = self.get_config().await?;
+        if let Some(product) = config.product {
+            Ok(product)
+        } else {
+            Ok("".to_owned())
+        }
+    }
+
+    /// Selects the product to install
+    pub async fn select_product(&self, product_id: &str) -> Result<(), ServiceError> {
+        let config = SoftwareConfig {
+            product: Some(product_id.to_owned()),
+            patterns: None,
+        };
+        self.set_config(&config).await
+    }
+
+    pub async fn get_registration(&self) -> Result<RegistrationInfo, ServiceError> {
+        self.client.get("/software/registration").await
+    }
+
+    /// registration code used to register product
+    pub async fn registration_code(&self) -> Result<String, ServiceError> {
+        let ri = self.get_registration().await?;
+        Ok(ri.key)
+    }
+
+    /// email used to register product
+    pub async fn email(&self) -> Result<String, ServiceError> {
```

Comment:
For consistency, I would use `registration_email`. Anyway, I would get rid of `registration_code` and `registration_email` completely. They are using in `ProductStore::load` and it causes the `GET /software/registration` to happen twice:

```rust
let registration_code = self.product_client.registration_code().await?;
let email = self.product_client.email().await?;
```